File: py/aoc2021/day3.py
from collections import Counter


def part1(lines):
    """
    >>> part1(["00100", "11110", "10110", "10111", "10101", "01111", "00111", "11100", "10000", "11001", "00010", "01010"])
    198
    """
    gamma = "".join(Counter(column).most_common(1)[0][0] for column in zip(*lines))
    epsilon = "".join(chr(ord(c) ^ 1) for c in gamma)
    return int(gamma, 2) * int(epsilon, 2)


def part2(lines):
    """
    >>> part2(["00100", "11110", "10110", "10111", "10101", "01111", "00111", "11100", "10000", "11001", "00010", "01010"])
    230
    """
    o2 = lines
    co2 = lines
    i = 0
    while len(o2) > 1 or len(co2) > 1:
        common = max(
            Counter(s[i] for s in o2).items(), key=lambda entry: (entry[1], entry[0])
        )[0]
        uncommon = min(
            Counter(s[i] for s in co2).items(), key=lambda entry: (entry[1], entry[0])
        )[0]
        o2 = [s for s in o2 if s[i] == common]
        co2 = [s for s in co2 if s[i] == uncommon]
        i += 1
    return int(o2[0], 2) * int(co2[0], 2)


parts = (part1, part2)

File: rs/src/day3.rs
use super::util::{self, transpose};
use std::char;
use std::collections::HashMap;
use std::error::Error;
use std::hash::Hash;

fn counter<I>(items: I) -> HashMap<I::Item, usize>
where
    I: Iterator,
    <I as Iterator>::Item: Eq,
    <I as Iterator>::Item: Hash,
{
    let mut counts = HashMap::new();
    for item in items {
        counts.entry(item).and_modify(|n| *n += 1).or_insert(1);
    }
    counts
}

pub fn part1<'a, I, S>(lines: I) -> Result<u32, Box<dyn Error + Send + Sync>>
where
    I: IntoIterator<Item = &'a S>,
    S: AsRef<str> + 'a,
{
    let lines = lines
        .into_iter()
        .map(|line| line.as_ref().to_string())
        .collect::<Vec<_>>();
    let (gamma, epsilon): (String, Vec<_>) = transpose(lines.iter().map(|line| line.chars()))
        .filter_map(|column| {
            Some(
                *counter(column.into_iter())
                    .iter()
                    .max_by_key(|(_, count)| *count)?
                    .0,
            )
        })
        .map(|char| (char, char::from_u32(u32::from(char) ^ 1)))
        .unzip();
    let epsilon = epsilon
        .into_iter()
        .collect::<Option<String>>()
        .ok_or(util::Error)?;
    Ok(u32::from_str_radix(&gamma, 2)? * u32::from_str_radix(&epsilon, 2)?)
}

pub fn part2<'a, I, S>(lines: I) -> Result<u32, Box<dyn Error + Send + Sync>>
where
    I: IntoIterator<Item = &'a S>,
    S: AsRef<str> + 'a,
{
    let mut o2 = lines
        .into_iter()
        .map(|line| line.as_ref().to_string())
        .collect::<Vec<_>>();
    let mut co2 = o2.clone();
    let mut i = 0;
    while o2.len() > 1 || co2.len() > 1 {
        let common = *counter(o2.iter().filter_map(|line| line.as_bytes().get(i)))
            .into_iter()
            .max_by_key(|(byte, count)| (*count, *byte))
            .ok_or(util::Error)?
            .0;
        let uncommon = *counter(co2.iter().filter_map(|line| line.as_bytes().get(i)))
            .into_iter()
            .min_by_key(|(byte, count)| (*count, *byte))
            .ok_or(util::Error)?
            .0;
        o2.retain(|line| line.as_bytes().get(i) == Some(&common));
        co2.retain(|line| line.as_bytes().get(i) == Some(&uncommon));
        i += 1;
    }
    Ok(u32::from_str_radix(o2.first().ok_or(util::Error)?, 2)?
        * u32::from_str_radix(co2.first().ok_or(util::Error)?, 2)?)
}

#[cfg(test)]
mod tests {
    use super::*;
    use pretty_assertions::assert_eq;

    static EXAMPLE: &[&str] = &[
        "00100", "11110", "10110", "10111", "10101", "01111", "00111", "11100", "10000", "11001",
        "00010", "01010",
    ];

    #[test]
    fn part1_examples() -> Result<(), Box<dyn Error + Send + Sync>> {
        assert_eq!(198, part1(EXAMPLE)?);
        Ok(())
    }

    #[test]
    fn part2_examples() -> Result<(), Box<dyn Error + Send + Sync>> {
        assert_eq!(230, part2(EXAMPLE)?);
        Ok(())
    }
}
